### ISSUE TITLE
Maps conflict depending on species tree topology

### ISSUE THREAD

**Issue by conJUSTover** [gwct/grampa#2]
Thanks for developing this great program! 

I'm currently working on an ancient allopolyploidy event in which we have both progenitor lineages sampled, as well as an outgroup. Since there are two possible bifurcating species trees, I tried to run your program to see if I would get the same results. Below is the summary output of both runs: 

#---------------
Species tree: (T,(U,(O,(G,B)<1>)<2>)<3>)<4>

The MUL-tree with the minimum parsimony score is MT-40:	(T,((U,(G*,B*)<1>)<2>,(O,(G+,B+)<3>)<4>)<5>)<6>
Score = 26138
Species tree node	# dups mapped
B	514
G	535
<3>	47
G*	397
O	1187
<4>	471
<1>	246
U	728
T	556
<2>	966
<6>	380
B*	549
<5>	1537
#-----------------

Species tree: (T,(O,(U,(G,B)<1>)<2>)<3>)<4>

The MUL-tree with the minimum parsimony score is MT-37: (T,((O,(G*,B*)<1>)<2>,(U,(G+,B+)<3>)<4>)<5>)<6>
Score = 26138
Species tree node       # dups mapped
B       801
G       695
<3>     253
G*      237
O       1187
<2>     471
<4>     983
<1>     39
U       728
T       556
<6>     380
B*      262
<5>     1526
#--------------

My question is, even though the overall topology of the MUL-tree is the same, the duplication events mapped to the MUL-tree is different depending on which starting species tree topology I used. Is there a correct way to interpret this? My goal is to try and identify sets of genes belonging to each sub genome, and parsing the mapped gene trees onto the MUL-tree seems like a promising method if I can understand what is causing this discordance. 

Thanks for all your help!

**Comment by gwct** [comment]
Hi Justin,

Without seeing more of the data, I think that this is happening because some gene trees might have multiple maps that tie for the lowest score on a given species tree. When this happens, GRAMPA is forced to pick one of the mappings. As you can see this doesn't affect the overall score, but it may affect on which nodes some of the duplications are reported. If you check the detailed output file (ending with _det.txt in the output directory) you might see lots of lines starting with an asterisk like this:

* GT-9 to MT-76	2 maps found!
* GT-9 to MT-76	1	4	5
* GT-9 to MT-76	1	4	5

The first line tells us that there were two mappings that tied for this gene tree/species tree combo, and the subsequent lines show us the number of duplications and losses for each map. 

If you haven't already, one thing you could try to explore this further is to use the --maps option. This will actually show you were the duplications are mapped for each gene tree/species tree combo in the detailed output file. If you scroll to the very bottom of the readme page (https://gwct.github.io/grampa/readme.html) there is a bit of explanation for this option.

Hopefully this helps a bit! Let me know if you have any more issues.

**Comment by conJUSTover** [comment]
Hi Gregg, 

Thanks for your quick response. I have run the --maps option, and I now see the gene trees that map several times to the species tree. Your explanation makes sense. 

For these gene trees that map more than one way to the MUL-tree, how are they incorporated into the final count of gene dups on the MUL-tree? I presume they are placed on the node present in the input species tree (rather than the one "created" by your program), but is it possible that every possible mappings is used? Are all ignored? 

Thanks again for your assistance!

**Comment by gwct** [comment]
When there are multiple maps from a gene tree to a species tree only one of them is used in the final count of duplications per node. Unfortunately in these cases there's no way to know which map is correct, so GRAMPA just picks one.

When I've run GRAMPA on several datasets I haven't noticed many gene tree/species tree combos that have multiple maps, so I thought this was an acceptable solution. However, I could see this being a problem on datasets with fewer taxa, making it more likely that more gene trees could have multiple maps. Then we might not be so confident in the final duplication counts.

Are you seeing a lot of gene trees with multiple maps? If so, it may be worth looking through those maps to see if there are any informative patterns (ie if its the same gene tree topology each time).
Also, would it be helpful if I added a statement to the main output that says how many gene trees have multiple maps?

**Comment by conJUSTover** [comment]
I think another complicating factor in my current dataset is that i believe I have two to three different allopolyploidy events (!) with relatively poor sampling of the maternal and paternal lineages...which is not the easiest  scenario to disentangle and leads to much confusing and conflicting results. 

I don't know that adding a statement at the main output would be much help for me specifically (but might be to other people who use this great program), but I believe adding a "consensus" tree in which the genes that have multiple maps are designated by a different symbol would be a useful addition, though I don't know how difficult this would be. I should also mention that I will be submitting this project for publication in the next few weeks, so please don't rush to get these changes for me. This would only be for people who use this program in the future. 

Another potential useful addition (again, this may be impossible based on your algorithm) is to allow the option to reconcile gene trees with a MUL-tree in which the duplicated clade is not monophyletic (if, for instance, you have multiple WGM events in your tree, you can account for all of them). Again, this may be a bit pie-in-the-sky thinking on my part and incredibly difficult to implement, but thought I would offer the suggestion. Please take it if you like, but also ignore if you like as well :) 

Thanks for all of your help. Keep doing great work, 
Justin

**Comment by gwct** [comment]
Hey Justin,

I went ahead and added a message to the main output that states the number of gene trees that have multiple maps and directs the user to the detailed output file for more info. Hopefully that will help in the future... thanks for bringing this to my attention!

As far as some sort of consensus tree I'm not sure how I would do that at this point. If I understand correctly what you're really after is some sort of 'cloudogram' of maps or something, which is more of an output visualization problem. That will be an interesting thing to look into!

For multiple WGDs, we showed with the wheat data that GRAMPA should be able to identify multiple events. Basically, we found that the two MUL trees that scored the lowest were the ones corresponding to the two WGDs and then the third lowest was the singly-labeled tree. So, given some good, error-free gene trees (if that's ever possible!), GRAMPA should find all WGD events before finally listing the singly-labeled tree. At least that's the idea!

Thanks for your feedback!
-Gregg
